Source/Task/AsyncLib.cpp

```diff
@@ -164,20 +176,42 @@ class AsyncBlockInternalGuard
 {
 public:
     AsyncBlockInternalGuard(_Inout_ AsyncBlock* asyncBlock) noexcept :
-        m_internal{ reinterpret_cast<AsyncBlockInternal*>(asyncBlock->internal) }
-    {
-        ASSERT(m_internal);
-        while (m_internal->lock.test_and_set()) {}
+        m_internal1{ reinterpret_cast<AsyncBlockInternal*>(asyncBlock->internal) },
+        m_internal2(nullptr)
+    {
+        ASSERT(m_internal1);
+        while (m_internal1->lock.test_and_set()) {}
+        
+        if (m_internal1->state != nullptr)
```

Comment:
>m_internal1->state  [](start = 12, length = 19)

should this be pinned to local var?  can't tell if this can change out from under you from another thread

Reply:
It shouldn't.  This is the only class that can mutate the state pointer and we grab the lock above.